google/oauth2/_client.py

```diff
@@ -59,7 +66,41 @@ def _handle_error_response(response_data):
     except (KeyError, ValueError):
         error_details = json.dumps(response_data)
 
-    raise exceptions.RefreshError(error_details, response_data)
+    raise exceptions.RefreshError(
+        error_details, response_data, retryable=retryable_error
+    )
+
+
+def _can_retry(status_code, response_data):
+    """Checks if a request can be retried by inspecting the status code
+    and response body of the request.
+
+    Args:
+        status_code (int): The response status code.
+        response_data (Mapping | str): The decoded response data.
+
+    Returns:
+      bool: True if the response is retryable. False otherwise.
+    """
+    try:
+        # For a failed response, response_body could be a string
+        error_desc = response_data.get("error_description") or ""
+        error_code = response_data.get("error") or ""
+
+        # Per Oauth 2.0 RFC https://www.rfc-editor.org/rfc/rfc6749.html#section-4.1.2.1
+        # This is needed because a redirect will not return a 500 status code.
+        retryable_error_descriptions = {"internal_failure", "server_error"}
```

Comment:
`internal_failure` was the existing condition. Not sure the context for it, but it is not in the RFC as you said.

`temporarily_unavailable` IMO we should not retry, as the server is trying to load shed or is overwhelmed.

Reply:
`temporarily_unavailable` is retryable I feel because it was of no fault from the user's request and it could succeed by retrying. If the server is receiving too many requests from this specific client/user and wants them to backoff, then the error code will be 429.
If server is overloaded due to cumulative requests from multiple clients, it can still return 429 to some requests to control the load. When 429 is returned we should not retry.
Since `temporarily_unavailable` comes for 503, I think we could retry.

Reply:
Ok, that sounds good to me. With the exponential backoff change I think it reasonable to assume that the impact on the servers will be reduced.

I'll make this change.

Reply:
This section of the RFC is related to 3LO right? its a redirect for code exchange. It is not supported by the library?

Reply:
I don't have the context for why this is included